File: usecases/mwaa-cognito-cdk/lib/mwaa-stack.ts
import { CfnOutput, StackProps, Stack } from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import * as mwaa from "aws-cdk-lib/aws-mwaa";
import { NagSuppressions } from "cdk-nag";
import { Construct } from "constructs";

interface DeploymentStackProps extends StackProps {
  readonly stage: string;
  readonly mwaaEnvironmentName: string;
  readonly mwaaPrivateSubnetIds: string[];
  readonly mwaaSecurityGroupIds: string[];
  readonly dagBucketArn: string;
  readonly dagBucketKmsKeyArn: string;
}

export class MwaaStack extends Stack {
  // Export value
  public readonly mwaaEnvironment: mwaa.CfnEnvironment;

  constructor(
    scope: Construct,
    id: string,
    readonly props: DeploymentStackProps,
  ) {
    super(scope, id, props);
    // Local configuration
    const MWAA_VERSION = "2.7.2";
    const MWAA_SIZE = "mw1.small";

    // Main MWAA Stack
    // IAM policy and role for MWAA services
    const mwaaPolicyDocument = new iam.PolicyDocument({
      statements: [
        new iam.PolicyStatement({
          actions: ["airflow:PublishMetrics"],
          resources: [
            `arn:aws:airflow:${this.region}:${this.account}:environment/*`,
          ],
        }),
        new iam.PolicyStatement({
          actions: ["s3:GetObject*", "s3:GetBucket*", "s3:List*"],
          resources: [
            `${this.props.dagBucketArn}`,
            `${this.props.dagBucketArn}/*`,
          ],
        }),
        new iam.PolicyStatement({
          actions: [
            "logs:CreateLogGroup",
            "logs:CreateLogStream",
            "logs:GetLogEvents",
            "logs:GetLogRecord",
            "logs:GetLogGroupFields",
            "logs:GetQueryResults",
            "logs:PutLogEvents",
          ],
          resources: [
            `arn:aws:logs:${this.region}:${this.account}:log-group:*`,
          ],
        }),
        new iam.PolicyStatement({
          actions: ["cloudwatch:PutMetricData", "logs:DescribeLogGroups"],
          resources: ["*"],
        }),
        new iam.PolicyStatement({
          actions: [
            "sqs:ChangeMessageVisibility",
            "sqs:GetQueueAttributes",
            "sqs:DeleteMessage",
            "sqs:GetQueueUrl",
            "sqs:ReceiveMessage",
            "sqs:SendMessage",
          ],
          resources: [`arn:aws:sqs:${this.region}:*:*`],
        }),
        new iam.PolicyStatement({
          actions: [
            "kms:Decrypt",
            "kms:DescribeKey",
            "kms:GenerateDataKey*",
            "kms:Encrypt",
            "kms:ReEncrypt*",
          ],
          resources: ["arn:aws:kms:*:*:key/*"],
        }),
      ],
    });
    const mwaaRole = new iam.Role(this, "MWAAExecutionRole", {
      assumedBy: new iam.CompositePrincipal(
        new iam.ServicePrincipal("airflow.amazonaws.com"),
        new iam.ServicePrincipal("airflow-env.amazonaws.com"),
      ),
      // custom description if desired
      description: "This is an execution role for MWAA services",
      inlinePolicies: { mwaa_policy: mwaaPolicyDocument },
    });

    // MWAA configuration
    // Logging configuration setup
    const mwaaLoggingConfigurationProperty: mwaa.CfnEnvironment.LoggingConfigurationProperty =
      {
        taskLogs: {
          enabled: true,
          logLevel: "INFO",
        },
        webserverLogs: {
          enabled: true,
          logLevel: "INFO",
        },
      };
    // Network configuration setup
    const mwaaNetworkConfigurationProperty: mwaa.CfnEnvironment.NetworkConfigurationProperty =
      {
        securityGroupIds: this.props.mwaaSecurityGroupIds,
        subnetIds: this.props.mwaaPrivateSubnetIds,
      };
    // MWAA cluster
    this.mwaaEnvironment = new mwaa.CfnEnvironment(this, "MWAA", {
      name: this.props.mwaaEnvironmentName,
      // Optional parameters
      airflowVersion: MWAA_VERSION,
      dagS3Path: "dags/",
      environmentClass: MWAA_SIZE,
      executionRoleArn: mwaaRole.roleArn,
      kmsKey: this.props.dagBucketKmsKeyArn,
      loggingConfiguration: mwaaLoggingConfigurationProperty,
      networkConfiguration: mwaaNetworkConfigurationProperty,
      sourceBucketArn: this.props.dagBucketArn,
      webserverAccessMode: "PRIVATE_ONLY",
    });

    // Stack Outputs
    new CfnOutput(this, "MWAARoleARN", {
      description: "MWAA role ARN",
      value: mwaaRole.roleArn,
    });
    new CfnOutput(this, "MWAAInternalUrl", {
      description: "MWAA internal access url",
      value: `https://${this.mwaaEnvironment.attrWebserverUrl}`,
    });

    // CDK NAG suppressions
    NagSuppressions.addResourceSuppressions(
      mwaaRole,
      [
        {
          id: "AwsSolutions-IAM5",
          reason: "False positive, the CDK generate this policies with *",
        },
      ],
      false,
    );
  }
}

File: usecases/mwaa-with-codeartifact/infra/mwaa_stack.py
import os

from aws_cdk import aws_mwaa as mwaa, aws_iam as iam, CfnJson, Stack
from constructs import Construct
from infra.s3_stack import S3Stack
from infra.vpc_stack import VpcStack


class MwaaStack(Stack):
    def __init__(
        self, scope: Construct, id: str, vpc: VpcStack, s3: S3Stack, **kwargs
    ) -> None:
        super().__init__(scope, id, **kwargs)

        mwaa_env_name = "mwaa_codeartifact_env"
        airflow_version = os.environ.get("AIRFLOW_VERSION", "2.0.2")

        # Create MWAA role
        mwaa_role = iam.Role(
            self,
            "MWAARole",
            assumed_by=iam.ServicePrincipal("airflow-env.amazonaws.com"),
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                resources=[
                    (
                        f"arn:aws:airflow:{self.region}:"
                        f"{self.account}:environment/{mwaa_env_name}"
                    )
                ],
                actions=["airflow:PublishMetrics"],
                effect=iam.Effect.ALLOW,
            )
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                resources=[
                    f"arn:aws:s3:::{s3.instance.bucket_name}",
                    f"arn:aws:s3:::{s3.instance.bucket_name}/*",
                ],
                actions=["s3:ListAllMyBuckets"],
                effect=iam.Effect.DENY,
            )
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                resources=[
                    f"arn:aws:s3:::{s3.instance.bucket_name}",
                    f"arn:aws:s3:::{s3.instance.bucket_name}/*",
                ],
                actions=["s3:*"],
                effect=iam.Effect.ALLOW,
            )
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                resources=[
                    (
                        f"arn:aws:logs:{self.region}:{self.account}:"
                        f"log-group:airflow-{mwaa_env_name}-*"
                    )
                ],
                actions=[
                    "logs:CreateLogStream",
                    "logs:CreateLogGroup",
                    "logs:PutLogEvents",
                    "logs:GetLogEvents",
                    "logs:GetLogRecord",
                    "logs:GetLogGroupFields",
                    "logs:GetQueryResults",
                    "logs:DescribeLogGroups",
                ],
                effect=iam.Effect.ALLOW,
            )
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                resources=["*"],
                actions=["cloudwatch:PutMetricData"],
                effect=iam.Effect.ALLOW,
            )
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                resources=[f"arn:aws:sqs:{self.region}:*:airflow-celery-*"],
                actions=[
                    "sqs:ChangeMessageVisibility",
                    "sqs:DeleteMessage",
                    "sqs:GetQueueAttributes",
                    "sqs:GetQueueUrl",
                    "sqs:ReceiveMessage",
                    "sqs:SendMessage",
                ],
                effect=iam.Effect.ALLOW,
            )
        )
        string_like = CfnJson(
            self,
            "ConditionJson",
            value={f"kms:ViaService": f"sqs.{self.region}.amazonaws.com"},
        )
        mwaa_role.add_to_policy(
            iam.PolicyStatement(
                not_resources=[f"arn:aws:kms:*:{self.account}:key/*"],
                actions=[
                    "kms:Decrypt",
                    "kms:DescribeKey",
                    "kms:GenerateDataKey*",
                    "kms:Encrypt",
                ],
                effect=iam.Effect.ALLOW,
                conditions={"StringLike": string_like},
            )
        )

        # Define MWAA configuration
        security_group_ids = mwaa.CfnEnvironment.NetworkConfigurationProperty(
            security_group_ids=[vpc.mwaa_sg.security_group_id],
            subnet_ids=vpc.get_vpc_private_subnet_ids,
        )

        logs_conf = mwaa.CfnEnvironment.ModuleLoggingConfigurationProperty(
            enabled=True, log_level="INFO"
        )

        logging_configuration = mwaa.CfnEnvironment.LoggingConfigurationProperty(
            scheduler_logs=logs_conf,
            webserver_logs=logs_conf,
            dag_processing_logs=logs_conf,
        )

        # Create MWAA environment
        mwaa_ca_env = mwaa.CfnEnvironment(
            self,
            "mwaa_ca_env",
            name=mwaa_env_name,
            environment_class="mw1.small",
            airflow_version=airflow_version,
            execution_role_arn=mwaa_role.role_arn,
            source_bucket_arn=s3.instance.bucket_arn,
            dag_s3_path="dags",
            requirements_s3_path="requirements.txt",
            max_workers=2,
            webserver_access_mode="PUBLIC_ONLY",
            network_configuration=security_group_ids,
            logging_configuration=logging_configuration,
        )
        mwaa_ca_env.node.add_dependency(mwaa_role)
